src/app/OperationalSessionSetup.cpp

```diff
@@ -191,27 +191,29 @@ void OperationalSessionSetup::UpdateDeviceData(const Transport::PeerAddress & ad
     if (mState == State::ResolvingAddress)
     {
         MoveToState(State::HasAddress);
-        err = EstablishConnection();
-        if (err != CHIP_NO_ERROR)
-        {
-            DequeueConnectionCallbacks(err);
-            // Do not touch `this` instance anymore; it has been destroyed in DequeueConnectionCallbacks.
-            return;
-        }
-    }
-    else
-    {
-        if (!mSecureSession)
+        mInitParams.sessionManager->UpdateAllSessionsPeerAddress(mPeerId, addr);
+        if (!mPerformingAddressUpdate)
         {
-            // Nothing needs to be done here.  It's not an error to not have a
-            // secureSession.  For one thing, we could have gotten a different
-            // UpdateAddress already and that caused connections to be torn down and
-            // whatnot.
+            err = EstablishConnection();
+            if (err != CHIP_NO_ERROR)
+            {
+                DequeueConnectionCallbacks(err);
+                // Do not touch `this` instance anymore; it has been destroyed in DequeueConnectionCallbacks.
+                return;
+            }
+            // We expect to get a callback via OnSessionEstablished or OnSessionEstablishmentError to continue
+            // the state machine forward.
             return;
         }
 
-        mSecureSession.Get().Value()->AsSecureSession()->SetPeerAddress(addr);
+        DequeueConnectionCallbacks(CHIP_NO_ERROR);
```

Comment:
Isn't this going to assert in `auto sessionHandle = mSecureSession.Get();` in that function because we won't have an associated session handle in the case where we're just performing an address update? Or am I missing something?

Reply:
Yeah in my follow up PR (that is not uploaded yet) I properly rename `auto sessionHandle = mSecureSession.Get();` to `auto optionalSessionHandle = mSecureSession.Get();` since `mSecureSession.Get()` returns `Optional<SessionHandle>`

The `sessionHandle.Value()` is not a problem since it does not get called since there are not callback associated with this.

Although I do agree since the only case we are expecting to be that we should just release to reduce that complexity and institutional knowledge about these subtle things. I will do that in the cleanup followup